lib/system-info.js

```diff
@@ -1,219 +1,296 @@
+    maybeAddHostUtilization(configHostname, utilizationConfig)
   }
 
-  let tasksDone = 0
-  function finishedResponse() {
-    if (++tasksDone === numTasks) {
-      callback(null, systemInfo)
-    }
+  if (Object.keys(utilizationConfig).length > 0) {
+    systemInfo.config = utilizationConfig
   }
 
-  module.exports._getProcessorStats(function getProcessCB(processorStats) {
-    systemInfo.packages = processorStats.packages
-    systemInfo.logicalProcessors = processorStats.logical
-    systemInfo.cores = processorStats.cores
-    finishedResponse()
-  })
-  module.exports._getMemoryStats(function getMemCB(memory) {
-    systemInfo.memory = memory
-    finishedResponse()
-  })
-  getKernelVersion(function getVersionCB(kernelVersion) {
-    systemInfo.kernelVersion = kernelVersion
-    finishedResponse()
-  })
-  utilization.getVendors(agent, function getVendorInfo(err, vendors) {
-    if (vendors) {
-      systemInfo.vendors = vendors
-    }
-    finishedResponse()
-  })
-  getBootId(agent, function reportBootId(err, bootId) {
-    if (bootId) {
-      systemInfo.bootId = bootId
-    }
-    finishedResponse()
-  })
+  const processorStats = await module.exports._getProcessorStats()
+  const memoryStats = await module.exports._getMemoryStats()
+  const kernelStats = await getKernelVersion()
+  const vendorStats = await getVendors(agent)
+  const bootId = await getBootId(agent)
+
+  maybeSetProcessorStats(processorStats, systemInfo)
+  maybeSetMemoryStats(memoryStats, systemInfo)
+  maybeSetKernelStats(kernelStats, systemInfo)
+  maybeSetVendorStats(vendorStats, systemInfo)
+  maybeSetBootId(bootId, systemInfo)
+
+  callback(null, systemInfo)
 }
 
-// placed on module for mocking purposes in tests
-module.exports._getProcessorStats = function getProcessorStats(callback) {
+/**
+ * Helper method for getting detailed, architecture specific processor information from the system
+ * Exported for testing purposes
+ *
+ * @returns {*} null if unknown platform, otherwise the processor stats
+ */
+module.exports._getProcessorStats = async function getProcessorStats() {
   const processorStats = {
     logical: null,
     cores: null,
     packages: null
   }
 
   if (platform.match(/darwin/i)) {
-    getSysctlValue(['hw.packages'], function getPackages(packages) {
-      getSysctlValue(['hw.physicalcpu_max', 'hw.physicalcpu'], function getCores(cores) {
-        getSysctlValue(
-          ['hw.logicalcpu_max', 'hw.logicalcpu', 'hw.ncpu'],
-          function getLogicalCpu(logical) {
-            processorStats.logical = parseFloat(logical, 10)
-            processorStats.cores = parseFloat(cores, 10)
-            processorStats.packages = parseFloat(packages, 10)
-
-            for (const key in processorStats) {
-              if (!processorStats[key] || !isInteger(processorStats[key])) {
-                processorStats[key] = null
-              }
-            }
-
-            callback(processorStats)
-          }
-        )
-      })
-    })
```

Comment:
(Applause gif)

Reply:
_this_ is why promises and then async/await was added to Node.js